txpool/txpool.go

```diff
@@ -306,27 +333,16 @@ func (t *TxPool) addImpl(origin TxOrigin, tx *types.Transaction) error {
 		return err
 	}
 
-	if t.slots+numSlots(tx) > t.maxSlots {
-		if !isLocal && t.Underpriced(tx) {
-			return ErrUnderpriced
-		}
-
-		dropped, success := t.Discard(t.slots-t.maxSlots+numSlots(tx), isLocal)
-		if !isLocal && !success {
-			return ErrTxPoolOverflow
-		}
-		for _, tx := range dropped {
-			mux := t.lockAccountQueue(tx.From, true)
-			if wrapper, ok := t.accountQueues[tx.From]; ok {
-				wrapper.accountQueue.Remove(tx.Hash)
-			}
-			mux.unlock()
-
-			t.pendingQueue.Delete(tx)
-			t.decreaseSlots(numSlots(tx))
-		}
+	// check for slot overflow and handle accordingly
+	t.gauge.Lock()
+	if err := t.processSlots(tx, isLocal); err != nil {
```

Comment:
Why don't you put the Lock & Unlock as well as increasing it by slots required inside the `processSlots` method itself? That way we have isolated all slots logic inside this method (which I think was the intention) and also we have eliminated the possibility of somebody else invoking the `processSlots` method without properly locking and unlocking around it.

Reply:
That's the way it was originally written a few commits back (https://github.com/0xPolygon/polygon-sdk/pull/213/commits/bab1157a912d15930dc152b2a0b5cf520b8ee9c7). But the `TestGaugeCheck` test showed on multiple runs that a race condition is still possible. 

This code fixes that at the expense of readability. And adds to the point of refactoring TxPool in general.

Reply:
[Refactored](https://github.com/0xPolygon/polygon-sdk/pull/213/commits/6a55bb60673518b5e8ff4cbb4a1c7d768e451999)